docs/source/notes/amp_examples.rst

```diff
@@ -0,0 +1,186 @@
+        scaler.step(optimizer)
+
+        # Update the scale for next iteration.
+        scaler.update()
+
+
+Scale-Aware Operations
+----------------------
+
+The gradients resulting from ``scaler.scale(loss).backward()`` are scaled.  ``scaler.step(optimizer)``
+automatically knows if it must unscale gradients before applying them.  However, networks that directly use
+gradients between the backward pass and the optimizer step should also be aware of the scale factor.
+
+Gradient clipping
+"""""""""""""""""
+
+Clipping is an operation that directly manipulates the gradients.
+:meth:`AmpScaler.unscale` may be used to unscale the gradients in-place before ``scaler.step(optimizer)``,
+allowing you to clip unscaled gradients as usual::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+        scaler.scale(loss).backward()
+
+        # Unscale the gradients owned by optimizer in-place
+        scaler.unscale(optimizer)
+
+        # Since the optimizer's owned gradients are unscaled, we can clip as usual:
+        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+
+        scaler.step(optimizer)
+        scaler.update()
+
+``scaler`` records that ``scaler.unscale(optimizer)`` has been called for this optimizer
+this iteration, so ``scaler.step(optimizer)`` knows not to redundantly unscale gradients before
+calling ``optimizer.step()``.
+
+Gradient penalty
+""""""""""""""""
+
+A gradient penalty loss term also requires awareness of the scale factor.
+Gradient penalty demonstrates the following use cases:
+
+* Creating out-of-place gradients with :func:`torch.autograd.grad`
+* Correct interaction of gradient scaling with double-backward
+
+Here's how that looks for a simple L2 penalty::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+
+        # We should scale outputs for the out-of-place backward pass
+        grad_params = torch.autograd.grad(scaler.scale(loss), model.parameters(), create_graph=True)
+
+        # In general, the penalty term may depend nonlinearly on the out-of-place gradients, so to be safe,
+        # manually unscale them before computing the penalty.  This unscale should be out-of-place and autograd-exposed.
+        grad_params = [p*(1./scaler.get_scale()) for p in grad_params]
+
+        # Compute the penalty term and add it to the loss
+        grad_norm = 0
+        for grad in grad_params:
+            grad_norm += grad.pow(2).sum()
+        grad_norm = grad_norm.sqrt()
+        loss = loss + grad_norm
+
+        # The usual scaling for backward will now accumulate leaf gradients that are appropriately scaled.
+        scaler.scale(loss).backward()
+        scaler.step(optimizer)
+        scaler.update()
+
+
+Gradient accumulation
```

Comment:
I think we can kill this section for now and revisit expanding the documentation when the with guard is added? I'd like to keep the PR as minimal as possible for now.